cmd/acra-translator/server/server.go

```diff
@@ -298,6 +312,29 @@ func (server *ReaderServer) Start(parentContext context.Context) {
 		}()
 	}
 	<-parentContext.Done()
+
+	// global 'cancel' has been called. Now we should wait (not more than specified duration) until all
+	// background goroutines spawned by readerServer will finish their execution
+	if timeoutExpired(&server.waitGroup, time.Second) {
+		log.Errorf("Couldn't stop all background goroutines spawned by readerServer. Exited by timeout")
+	}
+	return
+}
+
+// timeoutExpired waits for the waitgroup for the specified max timeout.
+// Returns true if waiting timed out.
+func timeoutExpired(wg *sync.WaitGroup, timeout time.Duration) bool {
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		wg.Wait()
+	}()
+	select {
+	case <-c:
+		return false // completed normally
+	case <-time.After(timeout):
+		return true // timed out
+	}
 }
```

Comment:
This neat piece of code is going to be useful in other places as well. I'd move it utilities, calling something like `WaitWithTimeout`, `TimedWait`, `TimeoutWait`, or something.

Reply:
Yes. But it is not very expressive naming for this function and I agree that it can be moved to our utility functions

Reply:
Done in https://github.com/cossacklabs/acra/pull/406/commits/55e177ef6709e55c8d819195c9989dda72a96d1f